src/codemodder/codetf.py

```diff
@@ -0,0 +1,137 @@
+"""
+Data models for the CodeTF format.
+
+We need to keep this in sync with the CodeTF schema.
+"""
+
+from __future__ import annotations
+
+import os
+import sys
+from enum import Enum
+from typing import TYPE_CHECKING, Optional
+
+from pydantic import BaseModel, model_validator
+
+from codemodder import __version__
+from codemodder.logging import logger
+
+if TYPE_CHECKING:
+    from codemodder.context import CodemodExecutionContext
+
+
+class Action(Enum):
+    ADD = "add"
+    REMOVE = "remove"
+
+
+class PackageResult(Enum):
+    COMPLETED = "completed"
+    FAILED = "failed"
+    SKIPPED = "skipped"
+
+
+class DiffSide(Enum):
+    LEFT = "left"
+    RIGHT = "right"
+
+
+class PackageAction(BaseModel):
+    action: Action
+    result: PackageResult
+    package: str
+
+
+class Change(BaseModel):
+    lineNumber: int
+    description: Optional[str]
+    # All of our changes are currently treated as additive, so it makes sense
+    # for the comments to appear on the RIGHT side of the split diff. Eventually we
+    # may want to differentiate between LEFT and RIGHT, but for now we'll just
+    # default to RIGHT.
+    diffSide: DiffSide = DiffSide.RIGHT
+    properties: Optional[dict] = None
+    packageActions: Optional[list[PackageAction]] = None
+
+
+class ChangeSet(BaseModel):
+    """A set of changes made to a file at `path`"""
+
+    path: str
+    diff: str
+    changes: list[Change] = []
+
+
+class Reference(BaseModel):
+    url: str
+    description: Optional[str] = None
+
+    @model_validator(mode="after")
+    def validate_description(self):
+        self.description = self.description or self.url
+        return self
+
+
+class Result(BaseModel):
+    codemod: str
+    summary: str
+    description: str
+    references: Optional[list[Reference]] = None
+    properties: Optional[dict] = None
+    failedFiles: Optional[list[str]] = None
+    changeset: list[ChangeSet]
```

Comment:
should this have `= []`?

Reply:
Hmm I don't think it's strictly necessary since this is a required field but it might be a good idea.